### ISSUE TITLE
Bug: \ref and \eqref return the wrong URL if an HTML base element is specified in the document's head section

### ISSUE THREAD

**Issue by fizzbowen** [mathjax/MathJax#1589]
### Issue Summary
Specifying an HTML `base` element in a document's `head` section may cause the `formatURL` function in the `equationNumbers` object to return an incorrect URL to the equation.

### Steps to Reproduce

1. Create an HTML document which specifies the URL of a different document (e.g., the host site's home page) as the `href` attribute of the `base` element in the `head` section.
2. Invoke the CDN version of MathJax via `<script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>` in the `head` section.
3. Use the default configuration, except enable auto-numbering by inserting `<script type="text/x-mathjax-config">MathJax.Hub.Config( { TeX: { equationNumbers: { autoNumber: "AMS"}}});</script>` into the `head` section.
4. Insert a numbered and labeled display equation into the document `body`, such as         `$$\begin{equation} \label{eq:general_linear_equation} {a_1}{x_1} + {a_2}{x_2} + {a_3}{x_3} + \ldots  + {a_n}{x_n} = b \end{equation}$$`.
5. Insert a reference to the equation elsewhere in the `body`, such as:
`A linear equation is "homogeneous" if and only if \(b = 0\) in Equation&nbsp;\( \eqref{eq:general_linear_equation}\).`
6. Upload the document to a Web server and open it in your favorite browser.
7. Mouse over the equation number (i.e., the "(1)") in the sentence where it is referenced to see the destination URL (or, if you are feeling brave, click the equation number to see where it takes you). Most likely, the destination URL will begin with the name of the document specified in the base element rather than the name of the document in which the equation appears.

This is a bug because...
...clicking on the referenced equation number should take you to that equation, not to an anchor in the base document.

Here is a fix which works for me (it restores the current document path to the URL)...
Locally changing the `formatURL` function in the `equationNumbers` object to read:
`formatURL: function (id) {return document.URL+'#'+escape(id)}`

However, I am not smart enough to know whether there would ever be a circumstance in which `document.URL` would also need to be escaped (as `id` currently is in the default version of `formatURL`), so this might be an additional consideration. And obviously making this the default would mess up references to an equation appearing in a different document.

### Technical details

* MathJax Version: master - latest commit:  Current CDN version as of Z 2016/08/20 19:00
* Client OS: Windows 10 Anniversary
* Browser: Chrome 52.0.2743.116 m, Firefox 48.0
*

**Comment by fizzbowen** [comment]
Updating my suggested fix to accommodate a greater variety of browser differences (the original one did not consistently work in Firefox following a page refresh, and the `.split` method is needed because Chrome's `window.location.host` returns the the hostname plus the trailing slash, whereas Firefox's omits the trailing slash):
`formatURL: function (id) {return window.location.protocol + '//' + window.location.host.split('/')[0] + window.location.pathname + '#' + escape(id)}`

**Comment by dpvc** [comment]
You are correct; if there is a `<base>` element, the references will not work properly.  One of the reasons for the `formatURL` function is so that you can customize it to your needs, as you are doing.  I assume you are using a `MathJax.Hub.Config()` call to make the change (rather than editing the MathJax source files themselves).

There are some issues (as you have found) with trying to include the original location as part of the line.  Another important case is when the page is generated by a server-side script and has "query parameters" (e.g., "file.html?query=...").  Your solution will fail in this case, and any solution may be problematic, since issuing the full URL may cause the page to be reloaded; if the page was generated as a response to a form being submitted (like a search form, or a login page), the needed data will not be part of the new URL, and may produce broken links.

It may be best to leave it in the hands of the page author who is setting the `<base>` element to configure MathJax to handle that situation properly by configuring `formatURL()` appropriately for his or her situation.

@pkra, we already needed to handle this situation in SVG output to get the links to the font cache to work as expected (though the solution has the same limitations as I describe here).  Should we do something similar for this case as well?  My recommendation would be to get the page URL in a separate variable (as we do in the SVG), making it be the empty string if there is no `<base>` element, and use `baseURL + '#' + escape(id)` as the result of `formatURL()`.  This will give the same results as the current version when there is no `<base>` (since `baseURL` is empty then), and will apply the best-guess of what the URL should be for when there *is* a `<base>` element.

**Comment by dpvc** [comment]
The `issue1589` branch of my fork of MathJax includes a work-around for the `<base>` issue.

**Comment by dpvc** [comment]
==> Merged

**Comment by dpvc** [comment]
We can't make a test for this in the current test framework, since including a `<base>` element will prevent the framework files from loading properly (since they use relative addresses).
